modules/packages/OrderedSet/Treap.chpl

```diff
@@ -0,0 +1,872 @@
+                      default-initialized, here: " + eltType: string);
+      }
+
+      _enter(); defer _leave();
+
+      var resultNode = _kth(_root, k);
+      if resultNode == nil {
+        var defaultValue: eltType;
+        return (false, defaultValue);
+      }
+      return (true, resultNode!.element);
+    }
+
+    /*
+      Returns the minimal element in the tree
+    */
+    pragma "no doc"
+    proc _first() {
+      var node = _root;
+      while node != nil && node!.children[0] != nil {
+        node = node!.children[0];
+      }
+      return node;
+    }
+
+    /*
+      Iterate over the elements of this orderedSet. Yields constant references
+      that cannot be modified.
+
+      .. warning::
+
+        Modifying this orderedSet while iterating over it may invalidate the
+        references returned by an iterator and is considered undefined
+        behavior.
+      
+      :yields: A constant reference to an element in this orderedSet.
+    */
+    iter const these() {
+      var node = _first();
+      while node != nil {
+        yield node!.element;
+        node = _neighbour(node, 1);
+      }
+    }
+
+    /*
+      Returns `true` if this orderedSet shares no elements in common with the orderedSet
+      `other`, and `false` otherwise.
+
+      :arg other: The orderedSet to compare against.
+      :return: Whether or not this orderedSet and `other` are disjoint.
+      :rtype: `bool`
+    */
+    proc const isDisjoint(const ref other: orderedSet(eltType, ?)): bool {
+      var result = true;
+
+      on this {
+        _enter(); defer _leave();
+
+        if !(size == 0 || other.size == 0) {
+
+          // TODO: Take locks on other?
+          for x in other do
+            if this._contains(x) {
+              result = false;
+              break;
+            }
+        }
+      }
+
+      return result;
+    }
+    
+    /*
+      Returns `true` if this orderedSet and `other` have at least one element in
+      common, and `false` otherwise.
+
+      :arg other: The orderedSet to compare against.
+      :return: Whether or not this orderedSet and `other` intersect.
+      :rtype: `bool`
+    */
+    proc const isIntersecting(const ref other: orderedSet(eltType, ?)): bool {
+      return !isDisjoint(other);
+    }
+
+    /*
+      Write the contents of this orderedSet to a channel.
+
+      :arg ch: A channel to write to.
+    */
+    proc const writeThis(ch: channel) throws {
+      _enter();
+      _visit(ch);
+      _leave();
+    }
+
+    /*
+      Returns `true` if this orderedSet contains zero elements.
+
+      :return: `true` if this orderedSet is empty.
+      :rtype: `bool`
+    */
+    inline proc const isEmpty(): bool {
+      var result = false;
+
+      on this {
+        _enter();
+        result = _root == nil;
+        _leave();
+      }
+
+      return result;
+    }
+
+    /*
+      Returns a new DefaultRectangular array containing a copy of each of the
```

Comment:
we can drop DefaultRectangular. It is not a user-facing term, and we use it to say "local" in many cases, "regular" in some other cases. If you want to specify you can say "regular local array" but even that is not needed IMHO.